### ISSUE TITLE
Incognito mode detected through fingerprintjs.com

### ISSUE THREAD

**Issue by uazo** [bromite/bromite#1615]
### Bromite version

Version: `95.0.4638.79`
Arch: `x86`
Android version: `10.0`
Device model: `emulator`

### Flags

None

### Is this bug about the SystemWebView?

No

### Is the bug reproducible with latest version?

Yes

### Can the bug be reproduced with corresponding Chromium version?

Yes

### Is this bug happening in an incognito tab?

Yes

### Allow ads for the site; is the bug caused by the adblocker?

No

### Is the bug a crash?

No

### Describe the bug

unfortunately the changes we made in https://github.com/bromite/bromite/issues/730 are no longer sufficient.

I think the time has come to remove this piece of code:

```
+//  if (is_incognito) {
+//    return CalculateIncognitoDynamicSettings(
+//        device_info_helper->AmountOfPhysicalMemory());
+//  }
```

### Steps to reproduce the bug

Steps to reproduce the bug:
1. Go to https://fingerprintjs.com in incognito

### Expected behavior

mitigate the possibility of being discovered.

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
I noticed that it is missing the template, please edit the issue to match the template for either a [bug](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) or a [feature request](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/feature_request.md); the issue will be automatically re-opened afterwards.

If this is not an issue nor a feature request and you want to start a discussion then you can use [GitHub Discussions](https://github.com/bromite/bromite/discussions).

**Tip**: if you use GitHub in Desktop mode the template will be proposed to you when submitting issues.

**Comment by csagan5** [comment]
@uazo please use the issue template, you removed `Is your feature request related to privacy?`

**Comment by uazo** [comment]
> you removed Is your feature request related to privacy

no, I opened it as a bug

I think it is due

> Can the bug be reproduced with corresponding Chromium version?
> Yes

however, it actually is.

**Comment by uazo** [comment]
there is also one thing.
reading the code it seems to me that there are other more or less simple ways to understand if the navigation is incognito.
try to solve them, as well as trying to understand if it is true (because I could be wrong and incorrectly think that they are exploitable), means making them public (because obviously before wasting time I would ask you for an opinion on the code), in your opinion what is better to do?

**Comment by csagan5** [comment]
There is nothing to keep secret: all the methods to detect incognito are already discovered by open source or closed source Javascript libraries. Keep it all in the open.

**Comment by uazo** [comment]
for this study I followed all the references to [BrowserContext::IsOffTheRecord()](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/browser_context.h;l=315;drc=57d99dad18948d6ba125d0f38f85f646e6793fc1;bpv=1;bpt=1) and it **seems** to me that there may be other ways to detect incognito mode.
probably, for each of these, a proof of concept should be done.

### detection via http headers and javascript
it seems that incognito only the first language is sent via the headers.

```
chrome/browser/net/profile_network_context_service.cc:
std::string ProfileNetworkContextService::ComputeAcceptLanguage() const {
  if (profile_->IsOffTheRecord()) {
    // In incognito mode return only the first language.
    return ComputeAcceptLanguageFromPref(
        language::GetFirstLanguage(pref_accept_language_.GetValue()));
  }
  return ComputeAcceptLanguageFromPref(pref_accept_language_.GetValue());
}
```
it should be noted that the default of the languages is taken from the string `IDS_ACCEPT_LANGUAGES` of the locales and there are always more than two.
for instance: https://source.chromium.org/chromium/chromium/src/+/main:components/strings/components_locale_settings_it.xtb

the same thing is detectable in javascript with `navigator.languages`

```
chrome/browser/renderer_preferences_util.cc:
std::string GetLanguageListForProfile(Profile* profile,
                                      const std::string& language_list) {
  if (profile->IsOffTheRecord()) {
    // In incognito mode return only the first language.
    return language::GetFirstLanguage(language_list);
  }
#if BUILDFLAG(IS_CHROMEOS_ASH)
  // On Chrome OS, if in demo mode, add the demo mode private language list.
  if (ash::DemoSession::IsDeviceInDemoMode()) {
    return language_list + "," + ash::DemoSession::GetAdditionalLanguageList();
  }
#endif
  return language_list;
}
```
possible solution: send only the defaults, whatever they are

### detection via subresource_redirect::LoginRobotsDeciderAgent

it's still not clear to me, but:
```
// The decider agent implementation that allows subresource redirect compression
// based on robots rules on non-logged-in pages. Currently only handles
// mainframes.

actually under kSubresourceRedirect disabled by default

// Enables redirecting subresources in the page to better compressed and
// optimized versions to provide data savings.
```
from [commit](https://chromium-review.googlesource.com/c/chromium/src/+/2764852)
in incognito is disabled:
```
content::BrowserContext*
LoginDetectionKeyedServiceFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  if (context->IsOffTheRecord())
    return nullptr;

  if (!IsLoginDetectionFeatureEnabled())
    return nullptr;

  return context;
}
```
rules are fetched from robots.txt (or from https://litepages.googlezip.net, I still don't understand)

### detection by checking the delay time of permission

idle detection and notification permission have wired a delay between 1 and 2 seconds before returning the block if incognito.
```
chrome/browser/idle/idle_detection_permission_context.cc
chrome/browser/notifications/notification_permission_context.cc

  // Idle detection permission is always denied in incognito. To prevent sites
  // from using that to detect whether incognito mode is active, we deny after a
  // random time delay, to simulate a user clicking a bubble/infobar. See also
  // ContentSettingsRegistry::Init, which marks idle detection as
  // INHERIT_IF_LESS_PERMISSIVE, and
  // PermissionMenuModel::PermissionMenuModel which prevents users from manually
  // allowing the permission.
  if (browser_context()->IsOffTheRecord()) {
    // Random number of seconds in the range [1.0, 2.0).
    double delay_seconds = 1.0 + 1.0 * base::RandDouble();
    VisibilityTimerTabHelper::CreateForWebContents(web_contents);
    VisibilityTimerTabHelper::FromWebContents(web_contents)
        ->PostTaskAfterVisibleDelay(
            FROM_HERE,
            base::BindOnce(&IdleDetectionPermissionContext::NotifyPermissionSet,
                           weak_factory_.GetWeakPtr(), id, requesting_origin,
                           embedding_origin, std::move(callback),
                           /*persist=*/true, CONTENT_SETTING_BLOCK,
                           /*is_one_time=*/false),
            base::Seconds(delay_seconds));
    return;
  }
```
multiple consecutive requests in incognito lead to a minimum delay of 1 second maximum 2, unlike normal navigation where the result of the promise is immediate after the user has granted (or not) access.

example:

```
IdleDetector.requestPermission().then((x) => { var d = new Date(); console.log(x + " " + d.getTime()) })
IdleDetector.requestPermission().then((x) => { var d = new Date(); console.log(x + " " + d.getTime()) })
IdleDetector.requestPermission().then((x) => { var d = new Date(); console.log(x + " " + d.getTime()) })
IdleDetector.requestPermission().then((x) => { var d = new Date(); console.log(x + " " + d.getTime()) })
```

non in incognito:
```
	denied 1638708970781
	denied 1638708970782
	denied 1638708970783
	denied 1638708970784
```

incognito:
```
	denied 1638709038935
	denied 1638709040163
	denied 1638709041894
	denied 1638709043425
```
possible solution: eliminate that ridiculous timeout

### detection via capture video manager
video capture is not possible in incognito.

```
chrome/browser/media/webrtc/desktop_capture_devices_util.cc
content/browser/media/capture_handle_manager.cc

media::mojom::CaptureHandlePtr CreateCaptureHandle(
...
  // Observing CaptureHandle wheneither the capturing or the captured party
  // is incognito is disallowed, except for self-capture.
  if (capturer->GetMainFrame() != captured->GetMainFrame()) {
    if (capturer->GetBrowserContext()->IsOffTheRecord() ||
        captured->GetBrowserContext()->IsOffTheRecord()) {
      return nullptr;
    }
  }
...
}
```
Try https://webrtc.github.io/samples/src/content/capture/video-video/ in normal and incognito mode.
note, to try in desktop, currently in android it doesn't seem to have effect.

possible solution: should the problem exist in bromite, the autoplay policy would inhibit that mechanism

### detection via open search
it seems that the html meta search related to the openearch are disabled in incognito mode:
```
chrome/browser/ui/search_engines/search_engine_tab_helper.cc
void SearchEngineTabHelper::PageHasOpenSearchDescriptionDocument(
    const GURL& page_url,
    const GURL& osdd_url) {
  // Checks to see if we should generate a keyword based on the OSDD, and if
  // necessary uses TemplateURLFetcher to download the OSDD and create a
  // keyword.
...
  Profile* profile =
      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
  if (page_url != web_contents()->GetLastCommittedURL() ||
      !TemplateURLFetcherFactory::GetForProfile(profile) ||
      profile->IsOffTheRecord())
    return;
...
  // Download the OpenSearch description document. If this is successful, a
  // new keyword will be created when done.
  TemplateURLFetcherFactory::GetForProfile(profile)->ScheduleDownload(
      keyword, osdd_url, entry->GetFavicon().url,
      frame->GetLastCommittedOrigin(), url_loader_factory.get(),
      frame->GetRoutingID(),
      content::GlobalRequestID::MakeBrowserInitiated().request_id);
}
```
but since otherwise a fetch is started at the address exposed by [OpenSearchDescriptionURL()](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/document.cc;l=6328;drc=8b58c961106e9f87da757b7dcb5d55dc5516ff2f;bpv=1;bpt=1), it is probably possible to understand if the mode is incognito.

possible solution: fetch and discard the result

I hope I am wrong and have not given new useful information to derive the incognito mode.

**Comment by uazo** [comment]
in addition it **seems** to me that there is a backdoor open only for google, surely without wanting to.
via
```
chrome/browser/sync/sync_encryption_keys_tab_helper.cc
chrome/renderer/sync_encryption_keys_extension.cc
```
and only for the url [https://accounts.google.com](https://source.chromium.org/chromium/chromium/src/+/main:google_apis/gaia/gaia_urls.cc;l=28;drc=8b58c961106e9f87da757b7dcb5d55dc5516ff2f;bpv=1;bpt=1) a new method is enabled in the `chrome` javascript object: `SetEncryptionKeys` which has a different behavior whether in anonymous mode or not.

**Comment by csagan5** [comment]
> I hope I am wrong and have not given new useful information to derive the incognito mode.

You could submit them to the Chromium issue tracker; perhaps there are also bounties? Probably not, you can check.



> in addition it **seems** to me that there is a backdoor open only for google, surely without wanting to. via
> 
> ```
> chrome/browser/sync/sync_encryption_keys_tab_helper.cc
> chrome/renderer/sync_encryption_keys_extension.cc
> ```
> 
> and only for the url [https://accounts.google.com](https://source.chromium.org/chromium/chromium/src/+/main:google_apis/gaia/gaia_urls.cc;l=28;drc=8b58c961106e9f87da757b7dcb5d55dc5516ff2f;bpv=1;bpt=1) a new method is enabled in the `chrome` javascript object: `SetEncryptionKeys` which has a different behavior whether in anonymous mode or not.

It is documented:
```
// API exposed by the browser process to the renderer process, such that the
// renderer can interact with the encryption state of Sync. Exposed to allowed
// origins only.
interface SyncEncryptionKeysExtension {
  // Provides sync encryption keys to the browser process. Returns true if no
  // further encryption keys are needed (which includes the case, among others,
  // of encryption keys not being needed even prior to this call).
  SetEncryptionKeys(string gaia_id, array<array<uint8>> encryption_keys,
                    int32 last_key_version) => ();
```

But I do not see this method available on the `chrome` Javascript object.

**Comment by uazo** [comment]
> You could submit them to the Chromium issue tracker; 

maybe this time I do, I'd be curious to see what they say.
do you agree to solve them as I indicated? it doesn't hurt, in any case.

> perhaps there are also bounties? Probably not, you can check.

for what i had read once, incognito mode bugs are out of any reward.

> But I do not see this method available on the chrome Javascript object.

because it is activated only for https://accounts.google.com
for that, since the sync in bromite with google is disabled, I would just remove the code from the gn.

**Comment by csagan5** [comment]
> do you agree to solve them as I indicated? it doesn't hurt, in any

It needs to be checked case by case:

> ### detection via http headers and javascript
> possible solution: send only the defaults, whatever they are

This change was introduced in https://bugs.chromium.org/p/chromium/issues/detail?id=1077547
Before reverting it one should understand why it was made; it looks like they introduced a bug there because sending the first locale was not the original intent; perhaps you can submit a bugfix upstream to correctly implement that.

> detection via subresource_redirect::LoginRobotsDeciderAgent

None of this applies to Bromite because there is no data saver integration; can you confirm?

> ### detection by checking the delay time of permission
> unlike normal navigation where the result of the promise is immediate after the user has granted (or not) access.
> possible solution: eliminate that ridiculous timeout

Also here, before changing you should check the rationale behind it:
* https://bugs.chromium.org/p/chromium/issues/detail?id=479679
* https://bugs.chromium.org/p/chromium/issues/detail?id=401439

You cannot detect this mechanism as you wrote because there is no guarantee that user denying a permission means incognito. We could tweak the randomness to be between something like 0.5 and 4 seconds but the mechanism is solid.

> ### detection via capture video manager
> video capture is not possible in incognito.

Equivalent to user denying it.

> ### detection via open search
> it seems that the html meta search related to the OpenSearch are disabled in incognito mode:

This goes back to the first commit with the Chromium source code, so we do not know why OpenSearch is not fetched in incognito mode; possibly because it is not supported.
I wonder what are the privacy problems with making it supported in incognito mode.

**Comment by uazo** [comment]
> Before reverting it one should understand why it was made; it looks like they introduced a bug there because sending the first locale was not the original intent
````
The current implementation --the one before, for us-- is causing lots Incognito 
language complaints and has an Incognito mode detector that needs to be fixed
for user privacy.  Chrome Privacy has asked for a merge.
````
````
Previously, when in Incognito the browser UI language was used for
Accept-Language and navigator.language.
This changes it so that only the first language in the Accept-Language
list is used.
````
I think they have already decided ...
in my opinion their assumption is that it does not give a 100% guarantee of the incognito mode, and they are right. probably have aggregate user data available as statistics ...
we can try to ask

> None of this applies to Bromite because there is no data saver integration; can you confirm?

I'm sorry, I don't know yet. that code is a bit convoluted and there are always active flags along with inactive flags, I have to study it.

> You cannot detect this mechanism as you wrote because there is no guarantee that user denying a permission means incognito.

actually yes, as long as the values of `Date.getTime()` as in my example are different by at least 1 second.
it doesn't matter if the user accept it or not, the response times matter.

> We could tweak the randomness to be between something like 0.5 and 4 seconds but the mechanism is solid.

nothing would change, at most we could just not respond to the promise, as if the user never made the decision.

> detection via capture video manager
> Equivalent to user denying it.

for bromite yes, there being the autoplay blocker, for the others I have some doubts.
how does the user block a video?

> detection via open search
> because it is not supported.

I don't know, I just need to check `chrome.csi()` to understand that it should be supported, I could try

**Comment by csagan5** [comment]
> > Before reverting it one should understand why it was made; it looks like they introduced a bug there because sending the first locale was not the original intent
> 
> ```
> The current implementation --the one before, for us-- is causing lots Incognito 
> language complaints and has an Incognito mode detector that needs to be fixed
> for user privacy.  Chrome Privacy has asked for a merge.
> ```
> 
> ```
> Previously, when in Incognito the browser UI language was used for
> Accept-Language and navigator.language.
> This changes it so that only the first language in the Accept-Language
> list is used.
> ```
> 
> I think they have already decided ... in my opinion their assumption is that it does not give a 100% guarantee of the incognito mode, and they are right. probably have aggregate user data available as statistics ... we can try to ask

Their rationale is about the user language leaking from the logged-in profile; since there is no such thing in Bromite we can make the accept-language the same for incognito and non-incognito.

> > You cannot detect this mechanism as you wrote because there is no guarantee that user denying a permission means incognito.
> 
> actually yes, as long as the values of `Date.getTime()` as in my example are different by at least 1 second. it doesn't matter if the user accept it or not, the response times matter.

You cannot: users can accept (or deny) permissions after any delay they like.

For example if you get this:
> non in incognito:
> 
> ```
> 	denied 1638708970781
> 	denied 1638708970785
> 	denied 1638708970792
> 	denied 1638708970804
> ```
> 
> incognito:
> 
> ```
> 	denied 1638709038935
> 	denied 1638709040163
> 	denied 1638709041894
> 	denied 1638709043425
> ```

How can you detect which one is incognito? The `denied` log message happens only when user clicks if not in incognito.


> > detection via capture video manager
> > Equivalent to user denying it.
> 
> for bromite yes, there being the autoplay blocker, for the others I have some doubts. how does the user block a video?

By not allowing the relative permission?

**Comment by uazo** [comment]
> Their rationale is about the user language leaking from the logged-in profile; since there is no such thing in Bromite we can make the accept-language the same for incognito and non-incognito.

you find me agree.

> For example if you get this:

assuming that:
```
	A1: 1638708970781
	A2: 1638708970785
	A3: 1638708970792
	A4: 1638708970804

	A1: 1638709038935
	A2: 1638709040163
	A3: 1638709041894
	A4: 1638709043425
```
I guess so: if `1 sec << A2-A1 << 2 sec` and `1 sec << A3-A2 << 2 sec` and `1 sec << A4-A3 << 2 sec` we note that they are incognito. the best solution is not to answer the request for permission.

perhaps the problem is reproducibility, considering that even with `Clear-Site-Data` http header the content settings are not deleted, but you could use a catch-all subdomain in an iframe and post the result to 1p via postMessage or other systems.

> detection via capture video manager
> Equivalent to user denying it.
> By not allowing the relative permission?

excuse my ignorance, but what permission are you referring to?

**Comment by uazo** [comment]
> I guess so: if 1 sec << A2-A1 << 2 sec and 1 sec << A3-A2 << 2 sec and 1 sec << A4-A3 << 2 sec we note that they are incognito. the best solution is not to answer the request for permission

no, it doesn't work like that, you are right.

**Comment by csagan5** [comment]
> > > detection via capture video manager
> > > Equivalent to user denying it.
> > 
> > 
> > for bromite yes, there being the autoplay blocker, for the others I have some doubts. how does the user block a video?
> 
> By not allowing the relative permission?

"capture video" permission: isn't this this the permission for the camera? Then user has to allow access to the camera.

**Comment by uazo** [comment]
>  isn't this this the permission for the camera?

no, from what I had seen (but we are talking about chromium desktop), even if that permission is not enabled the function is active.

**Comment by csagan5** [comment]
> > isn't this this the permission for the camera?
> 
> no, from what I had seen (but we are talking about chromium desktop), even if that permission is not enabled the function is active.

I probably need a diagram to understand this.

> > detection via open search
> > because it is not supported.
> 
> I don't know, I just need to check `chrome.csi()` to understand that it should be supported, I could try

I think we could allow this, in the end the OpenSearch feature is not doing anything special (in terms of privacy) when the search engine is added; it is completely passive.

**Comment by uazo** [comment]
> I probably need a diagram to understand this.

the problem does not exist in android as captureStream is not supported.
basically, with the desktop version, [captureStream](https://webrtc.github.io/samples/src/content/capture/video-video/) in  incognito doesn't (never) work, if instead you open it in normal mode, it works, regardless of permissions

> I think we could allow this,

if you agree
- an option in always incognito patch
- if deactivated, the call to the url is still made but the result is thrown away

in addition I would add the url of the template in the ui of the choice of the search engine

**Comment by csagan5** [comment]
> > I probably need a diagram to understand this.
> 
> the problem does not exist in android as captureStream is not supported. basically, with the desktop version, [captureStream](https://webrtc.github.io/samples/src/content/capture/video-video/) in incognito doesn't (never) work, if instead you open it in normal mode, it works, regardless of permissions

Ok, so this is a method to detect incognito that applies only to desktop; since Bromite is not for desktop, not on topic here.



> > I think we could allow this,
> 
> if you agree
> 
> * an option in always incognito patch
> * if deactivated, the call to the url is still made but the result is thrown away

Why should we throw it away? I am thinking that it should work in the same way in normal mode or incognito mode, and no option needed since there are no privacy concerns involved here. My assumption is that it is not enabled in Chromium because this is very old code and nobody cares, but there are no real reasons to have this functionality blocked in incognito.



> in addition I would add the url of the template in the ui of the choice of the search engine

So that the user can view what is being added?

**Comment by uazo** [comment]
> and no option needed since there are no privacy concerns involved here.

but it could be used to identify incognito mode

> but there are no real reasons to have this functionality blocked in incognito

I think it's not to leave any traces on the phone

> So that the user can view what is being added?

yes

**Comment by csagan5** [comment]
> > and no option needed since there are no privacy concerns involved here.
> 
> but it could be used to identify incognito mode

Can you explain how?

> > but there are no real reasons to have this functionality blocked in incognito
> 
> I think it's not to leave any traces on the phone

Adding a search engine is not a trace, it is an user choice; incognito mode is not "ephemeral mode". It is equivalent to adding a bookmark.

> > So that the user can view what is being added?
> 
> yes

Sounds like a good feature to have.

**Comment by uazo** [comment]
> Can you explain how?

no, I don't know yet, when I do I tell you.

> Adding a search engine is not a trace, it is an user choice; incognito mode is not "ephemeral mode". It is equivalent to adding a bookmark.

I have never used that function, I could be wrong, but it seems to me that the addition is automatic, while only the selection is manual

> Sounds like a good feature to have.

well

**Comment by csagan5** [comment]
> > Can you explain how?
> 
> no, I don't know yet, when I do I tell you.
> 
> > Adding a search engine is not a trace, it is an user choice; incognito mode is not "ephemeral mode". It is equivalent to adding a bookmark.
> 
> I have never used that function, I could be wrong, but it seems to me that the addition is automatic, while only the selection is manual

The automatic part is the fetching of the XML descriptor, only in normal mode; after removing the check in https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/search_engines/search_engine_tab_helper.cc;l=133 then normal mode and incognito mode will both fetch the descriptor and there is no way to detect incognito. This fetch already does not use cookies so it should be safe from that point of view.

> > Sounds like a good feature to have.
> 
> well

?

**Comment by csagan5** [comment]
@uazo I am tentatively changing the code so that it will allow adding search engines also from incognito mode; 

> in addition I would add the URL of the template in the UI of the choice of the search engine

Please let me know what you would like to do regarding this ^ and whether there are other actions here for this issue.

**Comment by csagan5** [comment]
> I think the time has come to remove this piece of code:
> 
> ```
> +//  if (is_incognito) {
> +//    return CalculateIncognitoDynamicSettings(
> +//        device_info_helper->AmountOfPhysicalMemory());
> +//  }
> ```

@uazo can you please verify if this change would be effective to stop detection of incognito mode? Otherwise we will have to close this issue and live with it.

**Comment by holequantum** [comment]
what is the status correctly no browser can block there fingerprinting i have tried with firefox and brave also.

**Comment by uazo** [comment]
> can you please verify if this change would be effective to stop detection of incognito mode

when I saw it long ago, yes, it was enough.
must be tested with v104, I will do it, but after the last of August (before I am not there)

**Comment by holequantum** [comment]
> > can you please verify if this change would be effective to stop detection of incognito mode
> 
> when I saw it long ago, yes, it was enough. must be tested with v104, I will do it, but after the last of August (before I am not there)

https://github.com/fingerprintjs/fingerprintjs
if this helps

**Comment by csagan5** [comment]
@uazo take your time, I will leave it assigned to you.

**Comment by uazo** [comment]
I checked and we need to remove

```
+//  if (is_incognito) {
+//    return CalculateIncognitoDynamicSettings(
+//        device_info_helper->AmountOfPhysicalMemory());
+//  }
```

I have the various patches ready, but they are above v105, I never know whether to wait for the corresponding version to be released. if you want I can give you the diffs of the changes only.

**Comment by csagan5** [comment]
Fixed in `105.0.5195.41`.
